### PR TITLE
Optimize threadpool implementation

### PR DESCRIPTION
**Pull request by zhaozg**
1. reuse cached lua_State for thread pool work scheduling
2. lua_State for thread pool can be gc

### CODE REVIEW

**Comment by zhaozg** [comment]
continue https://github.com/luvit/luv/pull/380

File: src/work.c [hunk 1]
```diff
@@ -1,4 +1,5 @@
 /*
+  i"close", luv_queue_close},
```
Comment (squeek502) [inline src/work.c]: ?

Looks like an accidental paste?

File: tests/test-work.lua [hunk 2]
```diff
@@ -15,11 +15,13 @@ return require('lib/tap')(function (test)
             assert(n*n==r)
             assert(#s==4096)
             if step < count then
-                _uv.queue_work(ctx,n,s)
+                if ctx then _uv.queue_work(ctx,n,s) end
```
Comment (squeek502) [inline tests/test-work.lua]: Why is this change needed? Does this imply a breaking change in how `work`/`threadpool` is used?

**Comment by zhaozg** [comment]
Done

File: tests/test-work.lua [hunk 3]
```diff
@@ -20,6 +20,8 @@ return require('lib/tap')(function (test)
                 if (step % 100==0) then
                     p(string.format('run %d%%', math.floor(step*100/count)))
                 end
+            else
+              ctx = nil
```
Comment (squeek502) [inline tests/test-work.lua]: Same question here. Is this now necessary?

Comment (zhaozg) [reply tests/test-work.lua]: Yes, need set ctx to nil.  because ctx -> callback -> ctx loop,  if ctx not nil,  ctx will not be gc.

Comment (squeek502) [reply tests/test-work.lua]: You're right, and this was always true (`__gc` didn't get called in this test in older versions of `luv` too without this `ctx = nil` change)
